File: src/main/java/com/tranhuudat/prms/service/ResourceAllocationService.java
package com.tranhuudat.prms.service;

import com.tranhuudat.prms.dto.BaseResponse;
import com.tranhuudat.prms.dto.resource_allocation.ResourceAllocationDto;
import com.tranhuudat.prms.dto.resource_allocation.ResourceAllocationSearchRequest;

import java.util.UUID;

public interface ResourceAllocationService {

    BaseResponse create(ResourceAllocationDto request);

    BaseResponse update(UUID id, ResourceAllocationDto request);

    BaseResponse delete(UUID id);

    BaseResponse getById(UUID id);

    BaseResponse getPage(ResourceAllocationSearchRequest request);

    /**
     * Xuất Excel theo template; chỉ lọc theo tháng của {@code request.getMonth()} (bắt buộc).
     *
     * @throws IllegalArgumentException nếu tháng null
     */
    byte[] exportResourceEmployeeExcel(ResourceAllocationSearchRequest request);
}

File: prms-web/src/app/pages/resource-allocation/services/resource-allocation.service.ts
import { Injectable } from '@angular/core';
import { BaseService } from '../../../core/services/base-service';
import { ResourceAllocationWritePayload } from '../models/resource-allocation.model';
import { ResourceAllocationSearchRequest } from '../models/resource-allocation-search.request';

@Injectable({
  providedIn: 'root',
})
export class ResourceAllocationService {
  private readonly apiUrl = '/api/v1/resource-allocations';

  constructor(private base: BaseService) {}

  getPage(request: ResourceAllocationSearchRequest) {
    return this.base.post(this.apiUrl + '/page', request);
  }

  getById(id: string) {
    return this.base.get(`${this.apiUrl}/${id}`);
  }

  create(request: ResourceAllocationWritePayload) {
    return this.base.post(this.apiUrl, request);
  }

  update(id: string, request: ResourceAllocationWritePayload) {
    return this.base.put(`${this.apiUrl}/${id}`, request);
  }

  delete(id: string) {
    return this.base.delete(`${this.apiUrl}/${id}`);
  }

  /** Xuất Excel theo tháng; backend chỉ áp dụng trường `month`. */
  exportExcel(request: { month: Date | string }) {
    return this.base.postBlob(`${this.apiUrl}/export/excel`, { month: request.month, voided: false });
  }
}
